proof/refine/ARM/IpcCancel_R.thy

```diff
@@ -2521,6 +2439,445 @@ lemma possibleSwitchTo_all_invs_but_ct_not_inQ':
+    apply (rule corres_guard_imp)
+      apply (rule corres_split[OF _ set_ep_corres])
+         apply (clarsimp simp del: fun_app_def not_None_eq K_bind_def)
+         apply (rule corres_split[OF rescheduleRequired_corres])
+           apply (erule cancel_all_ipc_corres_helper)
+          apply (rule_tac Q="?abs_guard list" in hoare_weaken_pre)
+           apply (rule hoare_strengthen_post)
+            apply (rule ball_mapM_x_scheme)
+              apply (intro hoare_vcg_conj_lift_left_fix; (solves \<open>wpsimp wp: gts_wp\<close>)?)
+               apply (wpsimp wp: restart_thread_if_no_fault_tcb_sts_of_other
+                                 reply_unlink_tcb_tcb_sts_of_other gts_wp)
+              apply (rule hoare_weaken_pre
+                     , rule cancel_all_ipc_loop_body_reply_unlink_ts_pred_other
+                     , fastforce)
+             apply (clarsimp simp: pred_conj_def)
+             apply (intro hoare_vcg_conj_lift_left_fix; (solves \<open>wpsimp wp: gts_wp\<close>)?)
+             apply (wpsimp wp: cancel_all_ipc_loop_body_valid_sched gts_wp)
+            apply simp
+           apply fastforce
+          apply simp
+         apply (rule_tac Q="?conc_guard list" in hoare_weaken_pre)
+          apply (rule hoare_strengthen_post)
+           apply (rule ball_mapM_x_scheme)
+             apply (intro hoare_vcg_conj_lift_left_fix; (solves wpsimp)?)
+             apply (rule hoare_weaken_pre, rule cancelAllIPC_loop_body_not_runnable'_other, simp)
+            apply (clarsimp simp: pred_conj_def)
+            apply (intro hoare_vcg_conj_lift_left_fix; (solves wpsimp)?)
+             apply (rule hoare_weaken_pre, rule cancelAllIPC_loop_body_weak_sch_act_wf, simp)
+            apply (wpsimp wp: cancelAllIPC_loop_body_valid_queues)
+           apply simp+
+        apply (simp add: ep_relation_def)
+       apply (wpsimp wp: hoare_vcg_all_lift hoare_vcg_const_Ball_lift hoare_vcg_imp_lift')+
+    done
+
+  show ?thesis
+    apply (simp add: cancel_all_ipc_def cancelAllIPC_def)
+    apply add_sym_refs
+    apply (rule corres_stateAssert_add_assertion[where P'=\<top>, simplified])
+    apply (rule corres_split'[OF _ _ get_simple_ko_sp get_ep_sp'])
+     apply (rule corres_guard_imp [OF get_ep_corres]
+            ; simp add: ep_relation_def get_ep_queue_def
+                   del: isNothing_def return_bind not_None_eq)
+     apply (rename_tac ep ep')
+     apply (case_tac "ep = Structures_A.endpoint.IdleEP")
+      apply (prop_tac "ep' = Structures_H.endpoint.IdleEP")
+       apply (simp add: ep_relation_def get_ep_queue_def)
+      apply clarsimp
+     apply (rule_tac ep=ep in IdleEP_versus_SendEP_and_RecvEP)
+      apply (simp add: ep_relation_def get_ep_queue_def, rename_tac q)
+     apply (subst not_IdleEP_case_split)
+      apply blast
+     apply (subst not_IdleEP_case_split')
+      apply (fastforce simp: ep_relation_def get_ep_queue_def)
```

Comment:
If you wanted both `case_split` lemmas to work with `simp` here (rather than needing `subst`), you would need assumptions in your context `ep =/= IdleEP` and `ep' =/= IdleEP`. You don't quite manage that here.

Firstly, your `case_tac` only generates evidence that `ep =/= IdleEP`. You could get it for `ep'` as well with something like this:
```
  apply (case_tac "ep = Structures_A.IdleEP \/ ep' = Structures_H.IdleEP")
   (* Just smash everything out in the trivial case. *)
   apply (case_tac ep; case_tac ep'; simp add: ep_relation_def get_ep_queue_def)
  (* The following clarsimp should produce both ep =/= IdleEP and ep' =/= IdleEP,
     and from that, I would hope the case_split rules would work as conditional simp rules. *)
  apply (clarsimp simp: not_IdleEP_case_split not_IdleEP_case_split')
```
That's just off the top of my head, so there's a fair chance it won't quite work. If it doesn't work, I'd be interested to know why.

I'm not saying that you need to change this. I'm just here to provoke some discussion about how to deal with annoying corners of these kinds of proofs. :-)

Reply:
This did work well, thanks! In using simp to perform the splitting, it will want to use `K_bind_def`, which messes up the helper lemma. So here I do `simp del: K_bind_def`. I think this is the right trade-off though; it's now only the only instance where I have to delete something from the simpset to get a helper lemma working, and the splitting is more straightforward